### ISSUE TITLE
w3m image preview flickering

### ISSUE THREAD

**Issue by taylorchu** [ranger/ranger#86]
the image shows in one blink. but xterm works fine.

**Comment by BwackNinja** [comment]
w3m was handling visuals and colormaps incorrectly. It always used the default rather than taking it from where it wants to draw to. Here's a patch for it:
https://gist.github.com/BwackNinja/60a344730170f9ce2163

**Comment by r8b7xy** [comment]
How to apply this patch?
Thanks in advance.

**Comment by warnaud** [comment]
Hi There,

I managed to get it working perfectly by setting:
set draw_borders true

In the ~/.config/ranger/rc.conf

Cheers!

**Comment by tats** [comment]
> w3m was handling visuals and colormaps incorrectly.
> https://gist.github.com/BwackNinja/60a344730170f9ce2163

@BwackNinja Thanks for the contribution to w3m.
Merged in the development repository for Debian's w3m.
cf. https://sourceforge.net/p/w3m/patches/72/

**Comment by fmthoma** [comment]
After applying @BwackNinja's patch I still had problems with resized images not being displayed correctly in urxvt with 32-bit enabled, showing semi-transparent artifacts on resized images. It looks like this is a bug (?) in imlib2, but can be worked around in w3m-img by using a different method for resizing the image. Now image preview in ranger works for me without `set draw_borders true`, even with background transparency enabled in urxvt.

Here is the patch, should be applied in addition to the one supplied by @BwackNinja: [https://gist.github.com/fmthoma/f76a1b44e00d5ca972bb](https://gist.github.com/fmthoma/f76a1b44e00d5ca972bb)

**Comment by tats** [comment]
> https://gist.github.com/fmthoma/f76a1b44e00d5ca972bb

@fmthoma Merged in Debian's w3m development repo.  Thank you.

**Comment by noctuid** [comment]
@tats @fmthoma Installing w3m from that repo does not fix image display in termite for me. It also breaks ranger's preview (but not using w3m directly) in both xterm and urxvt. Any ideas on why this happens?

**Comment by fmthoma** [comment]
Does it only break after my patch, or is it broken before? What library do you use for loading images, imlib or imlib2 or gkd-pixbuf? My patch applies only to the imlib2 backend.

I did not test with termite, but it works on xterm (unchanged by my patch, since xterm does not support 32-bit color anyways) and urxvt. In gnome-terminal, ranger's image preview does not work (I don't know why), but w3mimgdisplay works.

Have you tried running w3mimgdisplay separately?
`/usr/lib/w3m/w3mimgdisplay <<<$'0;1;100;100;100;100;;;;;/path/to/some/picture\n4;\n3;'`

**Comment by noctuid** [comment]
Installing from that repo, it's broken both before and after the patch. I tried using `--with-imagelib=imlib2` without success. I don't have `/usr/lib/w3m/w3mimgdisplay` (I guess I need to look through the Makefile), but moving the w3mimgdisplay binary there fixes the problem for xterm but not termite. I can view images in termite now using `w3m /path/to/image` without it opening an external image viewer, but I don's see anything (just the background color; all black).

**Comment by fmthoma** [comment]
Okay, I did not try to install from the development repo. I just fetch the sources (`apt-get source w3m`) and apply the patches. For Ubuntu (and probably other Debian based systems) have a look at my [dotfiles](https://github.com/fmthoma/config-files/blob/master/install/w3m).

**Comment by noctuid** [comment]
@fmthoma I'm not using a debian based distro. Maybe I'll try manually patching it myself. Have you tested the patch successfully with termite? It fixes the messed up previews in urxvt with a transparent background, but has no effect on termite for me (transparent or non-transparent background).

**Comment by fmthoma** [comment]
Hm, seems like I was a little too fast with my comment on https://github.com/thestinger/termite/issues/199.
If I have some spare time the next weeks, I'll have a look into it.

**Comment by fmthoma** [comment]
@noctuid: Today I had a go with termite. On my machine, w3mimgdisplay does work, but image display in ranger does not. This might be a curses/terminfo/termcap problem. You might want to open a new issue in the [ranger repo](https://github.com/hut/ranger) and start some debugging ([img_display.py](https://github.com/hut/ranger/blob/df64986fea91c9103baa1a07c5021cc0b4bb7a6f/ranger/ext/img_display.py) should be a good starting point, I suspect that one of the calls to w3mimgdisplay in this file fails in termite).

**Comment by noctuid** [comment]
@fmthoma I will mess around with it more, but I can't get w3mimgdisplay working at all with termite.

**Comment by Link-Satonaka** [comment]
Built the latest debian mirror git commit:
Under st:
- w3m <file.jpg> - doesn't work
- ranger preview doesn't work
- ranger full preview (press 'i') doesn't work

Under xterm:
- w3m <file.jpg> - doesn't work
- ranger preview works- But then again it already does with the standard w3m package...
- ranger full preview (press 'i') doesn't work

I applied the patch from @fmthoma 's dotfiles (one of them appeared to already be applied to the latest debian commit) but it did not change any of my results.

**Comment by sim590** [comment]
On archlinux, the package [seems to be using the debian repository](https://projects.archlinux.org/svntogit/packages.git/tree/trunk/PKGBUILD?h=packages/w3m)'s source. Using `urxvt` with the following configuration still produces the filckering bug.
```xdefaults
URxvt*font: xft:DejaVu Sans Mono:pixelsize=12:antialias=true:hinting=true
```
Using `xterm` or `terminator` works though.

**Comment by gavsiu** [comment]
@Link-Satonaka 

> Under xterm:
> - ranger full preview (press 'i') doesn't work

Also doesn't work under urxvt even with `set draw_borders true`

**Comment by webavant** [comment]
I have noticed that I can trigger the w3m image flicker issue in urxvt at will by viewing an image embedded local HTML document with two different doctypes, with XHTML Strict 1.0 preventing the flicker and HTML5 triggering the problem. Just some casual observations, in case it hasn't been noted before. I'm also noticing that ranger (in the same environment) is rendering images properly when it must perform scaling on the image to fit within the terminal, as if in a race where the image scaling procedure ultimately wins over the problem which caused the image to disappear.  In both cases noted, I still have flickering problems even with the successes, usually after scrolling (in w3m) or opening a menu (in ranger).

**Comment by copycat-killer** [comment]
I want to report my situation, in case it'd be useful for the devs.

Enabling the usual settings in `rc.conf`

```shell
set use_preview_script true
set preview_images true
```

and having an empty `scope.sh`:

```shell
#!/usr/bin/env sh

```

or making it not executable, the flickering/black bars issue disappears.

**Comment by Eizen** [comment]
> and having an empty scope.sh or making it not executable

^ Just wanted to confirm, this worked for me too. Though file `scope.sh` has to exist.

**Comment by Q3CPMA** [comment]
st user here,
same problem (sometimes, the image appear for a brief moment then dissapear). I also notice that it only happens when the image scaling/loading is too fast. Basically, there must be a race condition between w3m and the terminal redraw. Too bad, because the image preview could be really cool.

**Comment by nfnty** [comment]
urxvt image preview works with `preview_images_method urxvt` since `1.8.0`.

**Comment by copycat-killer** [comment]
@nfnty Yes, but requires `urxvt` to be built with `pixbuf` option, wich is not enabled in any standard rxvt-unicode package.

Also, the `urxvt` preview is still rubbing the image on the pane, instead of resizing it? If yes, then it's not preferable to `w3m`.

Anyway, `ranger` is still unable to render `w3m` previews without the spawning some black bands from time to time, and this happens not only with `urxvt`. This is still a bug.

**Comment by nfnty** [comment]
First test the `urxvt` preview and if that doesn't work properly or if you're having other issues with image preview and your issue hasn't already been reported, then open a new issue and fill out the template completely and also mention this issue. This issue thread is an incoherent mess.

**Comment by copycat-killer** [comment]
The `urxvt` preview works, anyway I prefer to use `w3m` because of the motivations I reported in my previous comment.

I recognize the thread is a mess, but it clearly reports this issue with at least the following terminals: `urxvt`, `st`, `termite`. Also, it has already been reported in different ways, for example: #185 #440 #472.

So it's very well known, but still not properly tackled IMO.

It'd be useful if you'd clarify the nature of this behaviour: is it related to `w3m` only, or `ranger` is culprit too, and how? Knowing this, we users would understand who we have to bother then. 😈

**Comment by nfnty** [comment]
You'll need to report this to `w3m` and your terminal emulator's devs. We can't do anything about it. If you find some new information then open a new issue.

**Comment by nfnty** [comment]
I'll just pool all the issues into this one, so you can continue the discussion here if you want, but don't expect us to resolve it.

**Comment by crocket** [comment]
Should we not replace w3m with something else or fork w3m?

**Comment by j-degreef** [comment]
This is clearly a problem between compton (or xcompmgr, maybe other composite managers) and w3mimgdisplay. It can be reproduced with w3m too so it's not a ranger problem.
I've spend hours testing this with all terminal manager I know of and all give the same result.
As the problem occurs with both compton and xcompmgr and only with w3mimgdisplay I think the problem resides inside the latest.
One workaround would be to resize the image before sending it to w3mimgdisplay but that's beyond Ranger task.
And I've not found any alternative to w3mimgdisplay...
I would be glad to help solving this problem but my coding knowledge is very limited.

**Comment by Link-Satonaka** [comment]
Some people have been using sixel programs successfully- and that is the way forward Imo.

**Comment by copycat-killer** [comment]
> This is clearly a problem between compton (or xcompmgr, maybe other composite managers) and w3mimgdisplay.

I can reproduce it without using a compositor or other window effects.

@Link-Satonaka What do you mean by "sixel" programs?

**Comment by Link-Satonaka** [comment]
Various implementations of libsixel.

Here's one example, there are many to choose from https://saitoha.github.io/libsixel/

**Comment by crocket** [comment]
I think https://github.com/saitoha/PySixel is a potent candidate for w3m.

**Comment by crocket** [comment]
I like sixel. I vote on SIXEL image preview.

**Comment by crocket** [comment]
Also, w3mimgdisplay is very slow when it tries to generate a preview thumbnail for any gif file larger than one megabyte. I have lots of large GIF files.

I might end up using `st` with image preview turned off.

**Comment by j-degreef** [comment]
> I can reproduce it without using a compositor or other window effects.
On what system and with what terminal ?
On Linux killing compton solved the problem on every test.

**Comment by crocket** [comment]
I'm a Gentoo Linux user.
I killed compton, and w3m flickers in `st-0.7.0`.

**Comment by copycat-killer** [comment]
> On what system and with what terminal ?

On Arch, with `urxvt`, `st` and `termite`.

**Comment by j-degreef** [comment]
@crocket @copycat-killer 
Are you talking about banding or flickering ?
Personally I was talking about banding, I've never faced flickering in my tests.

**Comment by copycat-killer** [comment]
@j-degreef I meant banding too.

**Comment by j-degreef** [comment]
@copycat-killer 
As we are both on Linux (Linux Mint 18.1 here) I don't understand, without compton (or any other comp manager) I have no problem. I cannot reproduce you problem on Termite and all VTE based TE.
There's no point using w3m with rxvt as it has it's own method (but bad font handling:-( )

With compton
![2017-04-26-204914_1366x768_scrot](https://cloud.githubusercontent.com/assets/6998555/25451469/8a8a0bd2-2ac2-11e7-89c2-6207731896c3.png)
Without compton
![2017-04-26-205034_1366x768_scrot](https://cloud.githubusercontent.com/assets/6998555/25451494/9d542e78-2ac2-11e7-990e-845c74770519.png)

**Comment by copycat-killer** [comment]
> There's no point using w3m with rxvt as it has it's own method (but bad font handling:-( )

The `urxvt` method is just rubbing the image on the terminal, without honoring the aspect ratio, plus it requires `urxvt` to be compiled with a certain non-standard option. Instead, `w3mimgdisplay` creates a thumbnail.

I don't use a background or transparency, so my banding "stripes" are black.

**Comment by Vifon** [comment]
> requires `urxvt` to be compiled with a certain non-standard option

Sorry for nitpicking but I'm aware of only one Linux distro that supplies `urxvt` without this option (Arch). I would rather call the **lack** of this option "non-standard" and file a bug report to the package maintainer.

As for the issue as a whole, not much can be done on our side. The way `w3m` previews work will always be buggy and it only gets worse with the additional stuff like compositors (e.g. `compton`).

> The urxvt method is just rubbing the image on the terminal, without honoring the aspect ratio

@copycat-killer, I cannot reproduce the aspect ratio issues. Could you provide a confirmed test case?
What do you mean by "rubbing"?

**Comment by crocket** [comment]
@Vifon rubbing means https://github.com/ranger/ranger/issues/815 which happens with urxvt image preview method. That's why I don't use urxvt image preview method.

**Comment by copycat-killer** [comment]
> Sorry for nitpicking but I'm aware of only one Linux distro that supplies urxvt without this option (Arch). I would rather call the lack of this option "non-standard" and file a bug report to the package maintainer.

It seems to me improbable that Arch is the only distro without it as standard, anyway there is `rxvt-unicode-pixbuf` in the AUR.

But enough discussing, let's focus on [PySixel](https://github.com/saitoha/PySixel). All we have to do is to create a `SIXELImageDisplayer` class which implement [this](https://github.com/ranger/ranger/blob/9572d80597aac9575615b488b894fd4d81490bd4/ranger/ext/img_display.py#L46-L59) interface.

**Comment by crocket** [comment]
I feel that https://github.com/saitoha/PySixel is unmaintained. Perhaps, there is a better sixel binding.

**Comment by j-degreef** [comment]
Sixel works on a limited number of terminals.
Unless people are ok to switch to another term, which I will not do for functionality reasons (Using i3 I need a term with keyboard navigation as Termite or urxvt have) that will not solve the problem for most.

**Comment by noctuid** [comment]
Maybe it's not an ideal solution for everyone, but tmux has much better keyboard navigation in my experience (and actually has customizable keybindings). I've not once used termite's keybindings. Regardless, I don't think limited terminal support is any reason not to add support for sixel as a method.

**Comment by copycat-killer** [comment]
Yeah, PySixel is unmaintained, and works on a limited number of terminals.

That's why there's [libsixel-python](https://github.com/saitoha/libsixel/tree/master/python).

As soon as I have enough time I'll try to implement the above mentioned displayer.

**Comment by Vifon** [comment]
Sixel support would certainly be nice if you could code it. It's a shame `urxvt` doesn't support it (personally I don't plan to switch terminals) but a new option is always welcome, as long as it won't be a hard dependency.

**Comment by crocket** [comment]
`st` will probably support SIXEL within this year because someone started pushing SIXEL support to `st`.
If he hears that ranger guys are going to add SIXEL support, he'll definitely push harder.

`st` is a viable alternative to urxvt and termite if you are willing to compile your configuration in C and accept XIM input method(which ibus supports) instead of the full ibus input method.

Gentoo makes it very easy to compile configurations in programming languages, so I use `st` along with `urxvt`.

**Comment by hobarrera** [comment]
> st is a viable alternative to urxvt and termite **if you are willing to compile your configuration from C** and accept XIM input method(which ibus supports) instead of the full ibus input method.

That's an absurdly huge `if`. I use a package manager to track that everything is up to date. I've no intention of ever changing that. A great deal of others thing the same. Manually staying up to date on releases and rebuilding is cumbersome and unmaintainable.

**Comment by crocket** [comment]
@hobarrera I am on gentoo linux. Gentoo's package manager makes it easy to compile st with a configuration file when it installs st. Thus, I can just reinstall the package after changing the configuration.

It takes less than 30 seconds to reinstall st on gentoo.

**Comment by Link-Satonaka** [comment]
You forgot to account for the amount of time it takes to fix your config to match upstream changes.

Even with gentoo, there is more management overhead. You and I are happy to deal with that, but respect that not everyone is.

**Comment by hobarrera** [comment]
> It seems st is not good on binary linux distributions. Perhaps, you want to check if it's possible to do the same on your linux distribution.

You also forgot that I need to subscribe to some for of RSS, and keep an eye on newer releases. Like @Link-Satonaka mentioned, also update the config every time upstream changes.
You might convince some people to do this, but 99% of the people prefer to have updates automated, rather than have to manually patch and install releases.

The idea of manually building from source (even if you build a package) will never be upstream because of the huge overhead. For those who prefer it: go ahead and use that, I've no issue. But don't expect the rest of the world to adopt that.

**Comment by crocket** [comment]
I agree. `st` better adopt a model similar to XMonad for configuration.
You can just install XMonad binary package which compiles haskell configuration file with GHC.
Unlike GHC, gcc, which `st` can use to compile configuration, is guaranteed to exist on almost every linux distribution.

**Comment by Link-Satonaka** [comment]
>By the way, ordinary text configuration files are just as susceptible to upstream changes

Only if the program[er] is not good (or doesn't particularly care about breaking-changes).

Your point restated in different words is "sometimes software is bad". Yes, yes it is :)

**Comment by steelvelveteen** [comment]
Hi, I got rid of the annoying black band across the previews by setting the border for ranger as true
Aprox around line 122 of the rc.conf file
set draw_borders true (it is false by default)
and this solved my problem
Hope it helps

**Comment by m3atatarian** [comment]
Just wanted to say that there is support for urxvt, but thus far it doesn't compile for me.

https://github.com/saitoha/rxvt-unicode-sixel

**Comment by slve** [comment]
The only workaround for the banding issue that worked for me is editing
`/usr/lib/python3.6/site-packages/ranger/ext/img_display.py`
adding `import time` at the top, 
and `time.sleep(1/100)` just right above `self.process.stdin.write(input_gen)`
around line 122

using 1/100 is reliable enough on my pc, still, not much lagging when scrolling through many images,
if you still see some bands on your pc, try increasing it,
using 1/20 not a single band showed up,
but scrolling through a lot of images wasn't as smooth as with 1/100

using this patch, there is no need set draw_borders true, which is a big plus for me

**Comment by wilzbach** [comment]
> The only workaround for the banding issue that worked for me is editing
> /usr/lib/python3.6/site-packages/ranger/ext/img_display.py
> adding import time at the top,
> and time.sleep(1/100)

Nice! That worked for me too. Is there any chance we could get such a workaround upstream?

**Comment by Vifon** [comment]
@wilzbach Already there. :) #1193

**Comment by doa379** [comment]
I'm experiencing this in Debian 10 Testing.
I was using w3m to preview images in Xterm and it's been working fine. But after switching to URxvt recently, images are flickering in the Ranger preview.
Clearly whatever fix that was pushed out has not really worked.

**Comment by doa379** [comment]
I would like to give credence to the theory that it's Xft related because the issue is not present when I start URxvt with a bitmap font instead of an Xft font.

**Comment by Ojousama** [comment]
![urxvt xft problem](https://0x0.st/zSPU.png)

**Comment by Ojousama** [comment]
![urxvt xft problem](https://0x0.st/zSPU.png "urxvt xft problem")

**Comment by DanielAndreasen** [comment]
I had the same flickering problem with `st` (and tried most things). What in the end worked for me, was to clone this repo and install from source (`sudo make install`). I made sure `ranger.py` was in my `PATH` (I made a link and called it `ranger`).
After that I tweaked the `w3m_delay` a bit and ended up with adding this to my config which seems to work in 95% of the cases
`set w3m_delay 0.1`

**Comment by amahmod** [comment]
adding `w3m_delay` to `rc.conf` solved my problem with `st`
Thanks to @DanielAndreasen

**Comment by thepenguinthatwants** [comment]
> adding `w3m_delay` to `rc.conf` solved my problem with `st`
> Thanks to @DanielAndreasen
Can you show your whole config?

I've being struggling with this for ages.
Pictures show up in ST or alacritty only milliseconds. urxvt and xterm they dont even show up.

set w3m_delay 0.1
didnt help it only delayed showing picture

```
set preview_files true
set preview_script ~/.config/ranger/scope.sh
set use_preview_script true
set preview_images_method w3m
set draw_borders true
```

**Comment by MarcoDiFrancesco** [comment]
> ```
> set draw_borders true
> ```

This solved the problem for me. Thanks!
